app/src/ui/delete-branch/delete-branch-dialog.tsx

```diff
@@ -8,31 +8,68 @@ import { ButtonGroup } from '../lib/button-group'
 import { Checkbox, CheckboxValue } from '../lib/checkbox'
 import { Dialog, DialogContent, DialogFooter } from '../dialog'
 import { Ref } from '../lib/ref'
+import { checkBranchExistsOnRemote } from '../../lib/git'
 
 interface IDeleteBranchProps {
   readonly dispatcher: Dispatcher
   readonly repository: Repository
   readonly branch: Branch
-  readonly existsOnRemote: boolean
   readonly onDismissed: () => void
 }
 
 interface IDeleteBranchState {
+  readonly existsOnRemote: boolean | null
   readonly includeRemoteBranch: boolean
 }
 
 export class DeleteBranch extends React.Component<
   IDeleteBranchProps,
   IDeleteBranchState
 > {
+  private checkingRemote = false
+
   public constructor(props: IDeleteBranchProps) {
     super(props)
 
     this.state = {
+      existsOnRemote: null,
       includeRemoteBranch: false,
     }
   }
 
+  private checkCompleted = () => {
+    this.checkingRemote = false
+  }
+
+  public componentWillReceiveProps(nextProps: IDeleteBranchProps) {
+    if (this.checkingRemote) {
+      // a check is being performed, don't start another
+      return
+    }
+
+    const upstreamChanged =
+      this.props.branch.upstream !== nextProps.branch.upstream
+
+    if (this.state.existsOnRemote == null || upstreamChanged) {
+      this.checkingRemote = true
+
+      checkBranchExistsOnRemote(this.props.repository, this.props.branch)
+        .catch(err => {
```

Comment:
Can we put the `then` callback before the `catch`?

Reply:
The `return false` line calls the `.then` handler with `false`, which lets the component pretend the branch doesn’t exist on the remote.

Reply:
@iAmWillShepherd yeah, that's what I had earlier but it resulted in duplicating the `setState` code, like this:

```ts
promise.then(existsOnRemote => {
   this.setState({ existsOnRemote }, this.checkCompleted)
}).catch(error => {
  log.error("oops", error)
  this.setState({ existsOnRemote }, this.checkCompleted)
})
```

By putting the `catch` first we can propagate a default value if the promise errors.

Reply:
Could you add that as a comment?

Reply:
9708fc8ddc02db4d8794b4fd87542384b05f5ddd